mne/dipole.py

```diff
@@ -138,11 +173,285 @@ def read_dipole(fname, verbose=None):
+    amp = np.array([r[1] for r in res])
+    ori = np.array([r[2] for r in res])
+    gof = np.array([r[3] for r in res]) * 100  # convert to percentage
+    return pos, amp, ori, gof
+
+
+def _fit_dipole(B, t, rrs, guess_fwd_svd, fwd_data, whitener):
+    """Fit a single bit of data"""
+    logger.info('---- Fitting : %7.1f ms' % (1000 * t,))
+
+    # make constraint function to keep the solver within the inner skull
+    if isinstance(fwd_data['inner_skull'], dict):  # bem
+        surf = fwd_data['inner_skull']
+
+        def constraint(rd):
+            if _points_outside_surface(rd[np.newaxis, :], surf, 1)[0]:
+                dist = _compute_nearest(surf['rr'], rd[np.newaxis, :],
+                                        return_dists=True)[1][0]
+                return -dist
+            else:
+                return 1.
+    else:  # sphere
+        R, r0 = fwd_data['inner_skull']
+        R_adj = R - 1e-5  # to be sure we don't hit the innermost surf
+
+        def constraint(rd):
+            return R_adj - np.sqrt(np.sum((rd - r0) ** 2))
+
+    # Find a good starting point (find_best_guess in C)
+    B2 = np.dot(B, B)
+    if B2 == 0:
+        logger.warning('Zero field found for time %s' % t)
+        return np.zeros(3), 0, np.zeros(3), 0
+    x0 = rrs[np.argmin([_fit_eval(rrs[fi][np.newaxis, :], B, B2, fwd_svd)
+                        for fi, fwd_svd in enumerate(guess_fwd_svd)])]
+    fun = partial(_fit_eval, B=B, B2=B2, fwd_data=fwd_data, whitener=whitener,
+                  constraint=constraint)
+
+    # Tested minimizers:
+    #    Simplex, BFGS, CG, COBYLA, L-BFGS-B, Powell, SLSQP, TNC
+    # Several were similar, but COBYLA won for having a handy constraint
+    # function we can use to ensure we stay inside the inner skull /
+    # smallest sphere
+    rd_final = optimize.fmin_cobyla(fun, x0, (constraint,), consargs=(),
+                                    rhobeg=5e-2, rhoend=1e-4, disp=False)
+
+    # Compute the dipole moment at the final point
+    Q, gof = _fit_Q(fwd_data, whitener, B, B2, rd_final)
+    amp = np.sqrt(np.sum(Q * Q))
+    norm = 1 if amp == 0 else amp
+    ori = Q / norm
+    return rd_final, amp, ori, gof
+
+
+@verbose
+def fit_dipole(evoked, cov, bem, mri=None, n_jobs=1, verbose=None):
+    """Fit a dipole
+
+    Parameters
+    ----------
+    evoked : instance of Evoked
+        The dataset to fit.
+    cov : str | instance of Covariance
+        The noise covariance.
+    bem : str | dict
+        The BEM filename (str) or a loaded sphere model (dict).
+    mri : str | None
+        The head<->MRI transform filename. Must be provided unless BEM
+        is a sphere model.
+    n_jobs : int
+        Number of jobs to run in parallel (used in field computation
+        and fitting).
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    dip : instance of Dipole
+        The dipole fits.
```

Comment:
I'd love to have an option to return the residual like we do it in the sparse solvers